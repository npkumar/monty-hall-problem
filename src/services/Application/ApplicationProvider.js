import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";

// Actions
const SET_SIMULATIONS = "SET_SIMULATIONS";
const SET_IS_CHANGE_DOOR = "SET_IS_CHANGE_DOOR";
const SET_WIN_PERCENTAGE = "SET_WIN_PERCENTAGE";
const SET_IS_SIMULATION_RUNNING = "SET_IS_SIMULATION_RUNNING";

export const ApplicationContext = createContext(null);

function reducer(state, action) {
  switch (action.type) {
    case SET_SIMULATIONS: {
      return { ...state, simulations: action.payload, winPercentage: null };
    }
    case SET_IS_CHANGE_DOOR: {
      return { ...state, isChangeDoor: action.payload, winPercentage: null };
    }
    case SET_WIN_PERCENTAGE: {
      return { ...state, winPercentage: action.payload };
    }
    case SET_IS_SIMULATION_RUNNING: {
      return { ...state, isSimulationRunning: action.payload };
    }
    default:
      throw new Error();
  }
}
const ApplicationProvider = ({ children }) => {
  const initialState = {
    simulations: 4000,
    isChangeDoor: true,
    winPercentage: null,
    isSimulationRunning: false,
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const value = {
    application: state,
    setSimulations: (simulations) =>
      dispatch({ type: SET_SIMULATIONS, payload: simulations }),
    setIsChangeDoor: (isChangeDoor) =>
      dispatch({ type: SET_IS_CHANGE_DOOR, payload: isChangeDoor }),
    setWinPercentage: (winPercentage) =>
      dispatch({ type: SET_WIN_PERCENTAGE, payload: winPercentage }),
    setIsSimulationRunning: (isSimulationRunning) =>
      dispatch({
        type: SET_IS_SIMULATION_RUNNING,
        payload: isSimulationRunning,
      }),
  };

  return (
    <ApplicationContext.Provider value={value}>
      {children}
    </ApplicationContext.Provider>
  );
};

ApplicationProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default ApplicationProvider;
