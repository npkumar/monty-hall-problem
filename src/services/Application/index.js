// Hooks
export { default as useApplication } from "./useApplication";

// Providers
export { default as ApplicationProvider } from "./ApplicationProvider";
