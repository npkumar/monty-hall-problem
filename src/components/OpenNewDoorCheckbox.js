import React from "react";
import { withStyles } from "@material-ui/core/styles";
import { green } from "@material-ui/core/colors";
import FormGroup from "@material-ui/core/FormGroup";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import Checkbox from "@material-ui/core/Checkbox";
import { useApplication } from "../services/Application";

const GreenCheckbox = withStyles({
  root: {
    color: "#52af77",
    "&$checked": {
      color: green[600],
    },
  },
})((props) => <Checkbox color="default" {...props} />);

export default function OpenNewDoorCheckbox() {
  const {
    application: { isSimulationRunning, isChangeDoor },
    setIsChangeDoor,
  } = useApplication();

  const handleChange = (event) => {
    setIsChangeDoor(event.target.checked);
  };

  return (
    <FormGroup row>
      <FormControlLabel
        data-testid="formControlLabel"
        control={
          <GreenCheckbox
            checked={isChangeDoor}
            onChange={handleChange}
            name="openagain"
            disabled={isSimulationRunning}
          />
        }
        label="Player chooses to change door"
      />
    </FormGroup>
  );
}
