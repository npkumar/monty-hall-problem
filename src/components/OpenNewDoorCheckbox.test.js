import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import { ApplicationContext } from "../services/Application/ApplicationProvider";
import OpenNewDoorCheckbox from "./OpenNewDoorCheckbox";

const value = {
  application: {
    application: { isSimulationRunning: false, isChangeDoor: true },
  },
};

describe("OpenNewDoorCheckbox", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const { asFragment } = render(
      <ApplicationContext.Provider value={value}>
        <OpenNewDoorCheckbox />
      </ApplicationContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have checkbox button", () => {
    const { getByTestId } = render(
      <ApplicationContext.Provider value={value}>
        <OpenNewDoorCheckbox />
      </ApplicationContext.Provider>
    );
    expect(getByTestId("formControlLabel")).toBeInTheDocument();
    expect(getByTestId("formControlLabel")).toHaveTextContent(
      "Player chooses to change door"
    );
  });
});
