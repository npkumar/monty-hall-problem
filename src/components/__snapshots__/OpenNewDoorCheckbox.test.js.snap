// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenNewDoorCheckbox renders 1`] = `
<DocumentFragment>
  <div
    class="MuiFormGroup-root MuiFormGroup-row"
  >
    <label
      class="MuiFormControlLabel-root"
      data-testid="formControlLabel"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root Component-root-1"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-5"
            data-indeterminate="false"
            name="openagain"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        Player chooses to change door
      </span>
    </label>
  </div>
</DocumentFragment>
`;
